File: Week4/Ex1.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

// #define m 1000;

int recus[1000000];

int power(int x, int y)
{
    if (y == 0)
    {
        recus[y] = 1;
        return recus[y];
    }
    if (recus[y] != 0)
    // tra ngay ket qua neu trong mảng recus da co luu ket qua tinh trước
    {
        return recus[y];
    }
    recus[y] = power(x, y - 1) * x % 1000;
    return recus[y];
}

int hashfunction(char *string)
{
    int len = strlen(string);
    int sum = 0;

    for (int i = 0; i < len; i++)
    {
        int coe = power(7, len - 1 - i);
        sum += string[i] * coe;
    }

    return sum % 1000;
}

typedef struct Node
{
    char string[50];
    struct Node *next;
} Node;

typedef struct HashTable
{
    Node *buckets[1000]; // == m
} HashTable;

void initHashTable(HashTable *ht)
{
    for (int i = 0; i < 1000; i++)
    {
        ht->buckets[i] = NULL;
    }
};

void insert(char *string, HashTable *ht)
{
    int key = hashfunction(string);
    Node *newNode = (Node *)malloc(sizeof(Node));
    strcpy(newNode->string, string);
    newNode->next = ht->buckets[key];
    ht->buckets[key] = newNode;
}

int find(char *string, HashTable *ht)
{
    int key = hashfunction(string);
    Node *tempNode = ht->buckets[key];
    while (tempNode != NULL)
    {
        if (strcmp(tempNode->string, string) == 0)
        {
            return 1;
        }
        tempNode = tempNode->next;
    }
    return 0;
}

int main()
{
    // 1st block
    char string[50];
    HashTable ht;
    initHashTable(&ht);
    while (strcmp(string, "*") != 0)
    {
        scanf("%s", string);
        insert(string, &ht);
    }
    char choice[10];
    int step[100000];
    int i = 0;
    do
    {
        scanf("%s", choice);
        if (strcmp(choice, "insert") == 0)
        {
            char test[50];
            scanf("%s", test);
            if (find(test, &ht) == 0)
            {
                insert(test, &ht);
                step[i] = 1;
                i++;
            }
            else
            {
                step[i] = 0;
                i++;
            }
        }
        else if (strcmp(choice, "find") == 0)
        {
            char test[50];
            scanf("%s", test);
            if (find(test, &ht) == 1)
            {
                step[i] = 1;
                i++;
            }
            else
            {
                step[i] = 0;
                i++;
            }
        }
    } while (strcmp(choice,"***") != 0);

    for(int j = 0; j < i; j++)
    {
        printf("%d\n",step[j]);
    }
}

File: Week4/Ex1.py
recus = [0] * 1000000

def power(x, y):
    if y == 0:
        recus[y] = 1
        return recus[y]
    if recus[y] != 0:
        return recus[y]
    recus[y] = power(x, y - 1) * x % 1000
    return recus[y]

def hashfunction(string):
    length = len(string)
    sum_val = 0

    for i in range(length):
        coe = power(7, length - 1 - i)
        sum_val += ord(string[i]) * coe

    return sum_val % 1000

class Node:
    def __init__(self):
        self.string = ""
        self.next = None

class HashTable:
    def __init__(self):
        self.buckets = [None] * 1000

def initHashTable(ht):
    for i in range(1000):
        ht.buckets[i] = None

def insert(string, ht):
    key = hashfunction(string)
    new_node = Node()
    new_node.string = string
    new_node.next = ht.buckets[key]
    ht.buckets[key] = new_node

def find(string, ht):
    key = hashfunction(string)
    temp_node = ht.buckets[key]
    while temp_node is not None:
        if temp_node.string == string:
            return 1
        temp_node = temp_node.next
    return 0

if __name__ == "__main__":
    string = ""
    ht = HashTable()
    initHashTable(ht)

    while string != "*":
        string = input()
        insert(string, ht)

    choice = ""
    step = []
    i = 0

    while choice != "***":
        choice = input().split()
        if choice[0] == "insert":
            test = choice[1]
            if find(test, ht) == 0:
                insert(test, ht)
                step.append(1)
                i += 1
            else:
                step.append(0)
                i += 1
        elif choice[0] == "find":
            test = choice[1]
            if find(test, ht) == 1:
                step.append(1)
                i += 1
            else:
                step.append(0)
                i += 1

    print(step)
